sgkit/stats/pedigree.py

```diff
@@ -976,3 +986,363 @@ def inverse_additive_relationships(
         }
     )
     return conditional_merge_datasets(ds, new_ds, merge)
+
+
+@numba_jit
+def _position_sort_pair(
+    pair: ArrayLike, position: ArrayLike
+) -> tuple:  # pragma: no cover
+    x, y = pair[0], pair[1]
+    if x < 0:
+        return (x, y)
+    elif y < 0:
+        return (y, x)
+    elif position[x] < position[y]:
+        return (x, y)
+    else:
+        return (y, x)
+
+
+@numba_jit
+def inbreeding_Hamilton_Kerr(
+    parent: ArrayLike,
+    tau: ArrayLike,
+    lambda_: ArrayLike,
+    allow_half_founders: bool = False,
+) -> ArrayLike:  # pragma: no cover
+    if parent.shape[1] != 2:
+        raise ValueError("Parent matrix must have shape (samples, 2)")
+    if not allow_half_founders:
+        _raise_on_half_founder(parent, tau)
+
+    n_samples = len(parent)
+    ploidy = tau.sum(axis=-1)
+    order = topological_argsort(parent)
+
+    # use a stack to track kinships that need calculating and add new dependencies
+    # to the top of the stack as they arise.
+    # the self-kinship of an individual depends on the kinship between its parents
+    # and (in the autopolyploid case) the self-kinship of each parent.
+    parental_self = np.unique(parent)
+    parental_self = np.broadcast_to(parental_self, (2, len(parental_self))).T
+    n_stack = n_samples + len(parental_self)
+    stack = np.empty((n_stack, 2), parent.dtype)
+    stack[0:n_samples] = parent[order]  # for kinship between pairs of parents
+    stack[n_samples:] = parental_self  # for self-kinship of each parent
+
+    # position of each sample within pedigree topology
+    position = np.empty(n_samples, dtype=np.int64)
+    for i in range(n_samples):
+        position[order[i]] = i
+
+    # calculate sparse kinship coefficients
+    kinship = dict()
+    idx = 0
+    while idx < n_stack:
```

Comment:
I'm finding it hard to grok what's happening in this loop. I know this is hard to do with an inherently complex algorithm like this, but I wonder if some of the bigger conditional cases could be split off into functions?